Fix issue with Hong Kong region normalization when paying with Apple Pay

reykjalin opened the pull request:
Fixes #1630 

# Changes proposed in this Pull Request:

- Instead of using the `billing_state` or `shipping_state` to normalize the region for a Hong Kong address, we instead use the `billing_postcode` and `shipping_postcode` which seem to contain the equivalent for Hong Kong addresses.

We've asked Apple whether these ^ observations about the address format are correct and are waiting for a reply. Until we've received a reply this PR will be labelled `pr: in progress`.

# Testing instructions

1. Add the following address to your Apple Pay shipping options:
    - Country: `Hong Kong`
    - Region: `New Territories`
    - District: `Taipo`
    - Address: `9 Feng Yuen Road`
2. `git checkout trunk`.
3. Add any product to your cart.
4. Attempt to checkout via Apple Pay using the address you just entered as a billing address and/or shipping address.
5. Notice that you are unable to checkout and no WooCommerce order is created.
6. `git checkout update-hong-kong-apple-pay-state-normalisation`
7. Try checking out via Apple Pay using the Hong Kong address.
8. Notice that you can now check out successfully.

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

Commit:
fix issue with HK region normalization

reykjalin commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -1359,12 +1370,12 @@ public function ajax_create_order() {
 			define( 'WOOCOMMERCE_CHECKOUT', true );
 		}
 
-		// In case the state is required, but is missing, add a more descriptive error notice.
-		$this->validate_state();
-
 		// Normalizes billing and shipping state values.
 		$this->normalize_state();
 
+		// In case the state is required, but is missing, add a more descriptive error notice.
+		$this->validate_state();
+
```
If we don't normalize before we validate the WooCommerce Checkout process will fail. I'm honestly not sure why.
`validate_state` doesn't interact with the `WC()->checkout()->process_checkout()` call AFAICT, but the normalization is ignored if we call `validate_state()` first. @ricardo do you have any ideas as to why?

ricardo replied:
> do you have any ideas as to why?

I reproduced the issue here and it's not that the normalization is ignored, but `validate_state` is throwing an error before the normalization happens, because `billing_state` and/or `shipping_state` are empty.

`validate_state` is used to display a friendlier error message for countries where the Payment Request API doesn't offer a state field, but WooCommerce expects it (so the country is not supported). The core team is working on a better approach for a more loose validation for Payment Request orders p7bje6-3fv-p2.

It's not really intuitive, but I think it's fine to call `normalize_state` before `validate_state` for this fix. Or do the Hong Kong mapping outside and before both methods.

I can't wrap my mind around "region" being passed as postcode from Apple Pay. To me it should be called "Postal code". It's really confusing to users 😪

ricardo replied:
On second thought. Apple Pay might not offering a state field, but a postal code field, much like many other countries.

I think we should confirm if Hong Kong buyers understand the keyword "Region" as postal code, or if this is some sort of i18n issue from Apple Pay for Safari in English, because for most countries it says "Postal code" in the field on that position.

In that case, we should be validating that field as postal code indeed, instead of implementing this change.

reykjalin replied:
I'm just waiting for the Apple Pay devs to set up a call with us so we can discuss this in more detail. Thanks for taking a look! I'll come back to this once I've had a chance to talk to Apple Pay 👍

Commit:
Merge branch 'develop' into update-hong-kong-apple-pay-state-normalisation

Commit:
fix linting issues

some assignments were aligned incorrectly

Commit:
Make region/state check more future proof

The change is intended to make sure the workaround won't break payments
after Apple Pay fix the bug this workaround is intended to address.

This involves a pretty exhaustive list of copy/pasted states using
https://www.rvd.gov.hk/doc/tc/hkpr13/06.pdf as a source.

Customers may still spell region names differently, in which case we'll
have to apply further normalization.

Commit:
Merge branch 'develop' into update-hong-kong-apple-pay-state-normalisation

ricardo reviewed (approved):
LGTM 💯

Should we add a changelog entry?

reykjalin commented:
> Should we add a changelog entry?

Ah, yes! I forgot that.

How does this sound:

```
Fix - Hong Kong addresses are now mapped more thoroughly to WooCommerce addresses when paying with Apple Pay.
```

I'm a bit wary about saying something along the lines of "all HK addresses now work with Apple Pay" because people might still write some of the districts using abbreviations that we're not handling correctly.

@ricardo let me know what you think! I'm happy to take suggestions on the phrasing too; I'm not sure what the best way to convey the nature of the fix is without making the entry too long or too technical 🤔

ricardo commented:
> I'm a bit wary about saying something along the lines of "all HK addresses now work with Apple Pay" because people might still write some of the districts using abbreviations that we're not handling correctly.

Yeah, that makes sense.

The suggested changelog looks good to me.

Commit:
Merge branch 'develop' into update-hong-kong-apple-pay-state-normalisation

Commit:
Add changelog entry

warmisland commented:
Just want to let you guys know that I'm still experiencing the same problem

EjayhanFernandes commented:
Issue still happening as reported here: 4925266-zd
